Raspberry Pi Camera driver

rohan5sep opened the pull request:
The submission includes customized libcamera stack packaged into a single libcamera.a library. The library was build to support following architectures -- 

1. ARMv7
2. ARMv8

The driver supports following Pi boards --

1. All the Pi boards supported by indigosky except Pi5 board -- I plan to include Pi5 support in coming month or two.

The driver does not support  MIPI CSI-2P formats for RAW data as of now. This can be included if required.

Driver is developed and tested with:
* Official Raspberry Pi High Quality Camera (IMX477R)
* Raspberry Pi Camera Module 3 NoIR (IMX708)
* Official Raspberry Pi Camera V2 (IMX219)
* OV9281-110 1MP Mono Camera for Raspberry Pi

rohan5sep commented:
Hi @rumengb could you please provide more details like RPi device and camera you used, call-stack, debugging messages etc?

rumengb commented:
Sure it is RPI 4B with 4GB RAM and RPI camera rev 1.3

It works fine with rpicam-xx tools:
![test](https://github.com/user-attachments/assets/b7cadbf0-9f81-4888-b40a-750d1244d0b3)

But indigo driver crashes every 2-3 frames and the acquired ones are broken:
![image](https://github.com/user-attachments/assets/50c5da6e-ff35-4158-a8a6-bf8c5d814670)

here is the log:
[indigo_rpi.log](https://github.com/user-attachments/files/20724604/indigo_rpi.log)

rumengb commented:
We also use Docker to build Indigo, and for RPi we use the following images:
"arm32v7/debian:bullseye-slim" 
"arm64v8/debian:bullseye-slim"
But honestly I doubt it is the image being used. Valgrind reports 17K invalid memory reads/writes.

Also, why did you add the compiled library in the source tree? Why do not you add the source and compile it on the fly as we do for Nexstar, FLI, Apogee and some other devices etc. We add only closed source SDKs as binaries.

kbingham commented:
libcamera's purpose is to abstract the camera API such that different platforms operate the same, so I expect more SBCs to be supportable.*

My reason to comment is that the work you are doing is not limited to Raspberry Pi - so the name of the plugin could be more descriptive.

*Source: me. libcamera - release manager.

I'm also curious why you need static linkage ? libcamera is fully open source - and doesn't need to be distributed as a binary.

kbingham commented on indigo_linux_drivers/ccd_rpi/Makefile.inc:
```diff
@@ -0,0 +1,30 @@
+
+SDK_INCLUDE ?= $(addprefix -I, $(wildcard bin_externals/*/include))
+SDK_LIB = -L$(SDK_DIR)
+SDK_LIB_A = $(SDK_DIR)/libcamera.a
```
What version of libcamera have you imported ? By pulling the whole sources in - it's not clear what changes you have made to the libcamera project on top ...

There must be a cleaner way to express this - or build it as a component and represent any patches you have on top ?

rohan5sep commented:
Point taken...let us see what indigo maintainers/owners think of...

rumengb commented:
If the SDK or library we use is open source and does not violate the license, we include the full source and, if possible, build it as a static library as part of our build process. As I can see, libcamera is BSD licensed, which is totally okay with the Indigo license. So, I do not see a problem with using this approach.

The approach of using the system libcamera is not very good for several reasons - sometimes we patch the SDKs to fix bugs. Sometimes vendor/author/distribution updates fix bugs, sometimes introduce others. And sometimes they are plain broken. So we use only the tested and "approved" versions of the libraries. Sometimes we keep some SDKs a version or two back just because of some regression that affects some users. And this happens more often than you think :)

So please use some version of the libcamera and add the source to /indigo_linux_drivers/ccd_rpi/externals/libcamera and build it as we do for some other SDK. I already gave you examples to look at in one of my previous posts here - https://github.com/indigo-astronomy/indigo/pull/652#issuecomment-2973059855

kbingham commented:
Just to clarify, libcamera is L-GPL. I don't know how that impacts the legalities of your SDK though. 

I suspect as long as it's possible to update the library, and it's clear what modifications are made to the code it's ok.

rohan5sep commented:
@kbingham is there a version of libcamera that builds and runs on 32-but bullseye os for Arch: ARMv7-A?

rohan5sep commented:
> 1. Libcamera is multilicense:
> 
> https://github.com/raspberrypi/libcamera/tree/main/LICENSES
> 
> among them 2 and 3 clause BSD and MIT... So no problem to add the source in indigo tree. 
> 
> 
> 
> 2. Also i would prefer to use this fork:
> 
> https://github.com/raspberrypi/libcamera/tree/main 
> 
> 
> 
> 3. This is the official rpi fork and they control the updates:
> 
> https://www.raspberrypi.com/documentation/computers/camera_software.html

Github one does not support 32-bit bullseye -- arm version as far as i understand. I used this fork to create libcamera.a

rohan5sep commented:
> You have to have the correct kernel drivers enabled.
> 
> And then there was probably a phase where raspberry pi kernel and libcamera versions were tied together.
> 
> But there's no differentiation between 32 bit or 64 bit. Both builds should work with the correct kernel drivers, which are provided by the raspberry pi bsp kernel.

Per this [link](https://libcamera.org/getting-started.html) the minimum meson compiler version is 0.63. This version of compiler is not supported on 32-bit bullseye.

kbingham commented:
> 1. Libcamera is multilicense:
>        https://github.com/raspberrypi/libcamera/tree/main/LICENSES
>        among them 2 and 3 clause BSD and MIT... So no problem to add the source in indigo tree.


It's interesting to see that this is causing confusion.

No - libcamera is not multi-licence. It's MANY licence.

libcamera itself is L-GPL. that means the core library (libcamera.so, libcamera.a) is explicitly L-GPL.

However - there are other components as part of libcamera which use other licenses such as BSD or MIT ... those include the IPA modules. How are you handling those in your static linking? It's essential that the IPA modules are kept aligned with the libcamera build.

libcamera IPA modules are the components that perform the 3a (auto-exposure, auto-whitebalance etc) ... those are separated and can be a different licence agreement. (In the case of Raspberry Pi - I think that's BSD which is why the BSD licence text is available).

>     2. Also i would prefer to use this fork:
>        https://github.com/raspberrypi/libcamera/tree/main
> 
>     3. This is the official rpi fork and they control the updates:
>        https://www.raspberrypi.com/documentation/computers/camera_software.html

Yes, if you want to run on a Raspberry Pi 5 - you will currently need to use the Raspberry Pi fork *and* the Raspberry Pi kernel accordingly.

rohan5sep commented:
So if I understand correctly, I am getting crashes in 32-bit version likely because the tool-chain I used to build the driver did not have Raspberry Pi kernel? I wonder why 64-bit libcamera.a is working fine though...

rohan5sep commented:
If instead of static linking we want to link libcamera the usual way then is there an option to copy libcamera directory having shared libraries and data to indigo/build/lib? The reason I am asking is that I do not see any directory structure there just a bunch of .so files.

libcamera build and data directories are as follows -- 
libcamera/src/ipa/rpi/pisp/data
libcamera/build/src/libcamera
libcamera/build/src/ipa
libcamera/build/src/ipa/libipa

rohan5sep commented:
> Please also remember that libcamera is more than just a single library file. It needs to link in the IPA modules at runtime _and_ read in the correct camera tuning files at runtime.
> 
> All of these locations can be customised - but you can not just move the files and expect them to work - you would have to make sure that the file locations are configured at compile time.

I am planning to set following meson flags at the setup stage -- datadir, datadir, libexecdir and sysconfdir. Or I could simply hard code path in the code in IPAManager::IPAManager().

kbingham commented:
> For static linking I converted *.json file to *.h and linked :)

Ayeeeee! I mean you can but aeeeee

rohan5sep commented:
> Probably just as well to set a custom prefix and keep the path's all consistent with a standard install but under your project specific prefix....

You mean set LIBCAMERA_IPA_MODULE_PATH along with Meson flags?

kbingham commented:
meson setup -Dprefix=/mysystem/library/path/

Then everything is relative to that path.

rohan5sep commented:
Any runtime dependencies will have to be added as well

rohan5sep commented:
@rumengb I have added the source code to the SDK. I have made changes to the original libcamera code to facilitate static linking. The code is also uploaded to libcamera fork [here](https://github.com/rohan5sep/libcamera.git). Removed 32-bit OS flags so the driver will only build on 64-bit OS.

rumengb commented:
Probably you should create a complete Makefile. Optional drivers are not built by the main makefile. See how it is done for andor.

rohan5sep commented:
Hi @rumengb the changes are done...let me know if something is missing/not done correctly...

rumengb reviewed (changes requested):
Once you address my comments, can you please drop this PR, sqash your 16 commits in this PR and make another with a single commit. As it is now I can not review the changes. It also fills the git history and uses space for the unused blobs.

rohan5sep commented on indigo_optional_drivers/ccd_rpi/indigo_ccd_rpi.cpp:
```diff
@@ -192,28 +192,41 @@ using ImageDimensions = std::vector<libcamera::Size>;
 using FrameInfo = std::map<libcamera::Size, libcamera::Size>;
 
 struct Logger {
```
Replaced logger_ with INDIGO_DRIVER_ERROR(...) everywhere. Removed Logger class.

rohan5sep commented on indigo_optional_drivers/ccd_rpi/Makefile:
```diff
@@ -0,0 +1,68 @@
+# RPi Camera Driver Makefile
+# Copyright (c) 2018 Rumen G. Bogdanovski
```
Done...

rohan5sep commented:
I have resolved the comments. I will now close this PR and reopen new one as requested...
